From patchwork Thu Dec 13 00:10:26 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10727395
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3C56E13AF
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 00:17:50 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2D0022B1A4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 00:17:50 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 213CB2B1A8; Thu, 13 Dec 2018 00:17:50 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C38A22B1A4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 13 Dec 2018 00:17:49 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728537AbeLMARs (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 12 Dec 2018 19:17:48 -0500
Received: from mail-qt1-f202.google.com ([209.85.160.202]:54587 "EHLO
        mail-qt1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726278AbeLMARs (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 12 Dec 2018 19:17:48 -0500
Received: by mail-qt1-f202.google.com with SMTP id z6so253472qtj.21
        for <kvm@vger.kernel.org>; Wed, 12 Dec 2018 16:17:47 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=I7YbaCkuvWQWVxzdePwFfJ7i8vJxzuwCxlMiX/NFYdw=;
        b=Y/bEC4HPS0sudou5zjLtBX44EdeYH1exGFc9906MAC5N+af1xO9ePHduCEEMbBRe4/
         ShlCmHjmCoV3R6+VS9BdyjpbizpHUtf9FK9bT55zNZzKGee6pG2PLEztAKHCc7gdU+7I
         mvMWPhbyQuYgcwavPP3+w7JNgWLDOZEyE4BdrSqgT1746UQFv0kHOi7GTqo87gRfz9Is
         RPxwhSthoiS9ghNKzl1BPpp1BGeFxgEhhH3Bv3/yS54h4Luo7WZe5pomZc4R5zBy5iUz
         MLGtY6QUmEbx//9YFKej87rb0AZzDWGgZocwkEVKqFBeyglEh2euNop68QRZqJw/LJbS
         Rz3w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=I7YbaCkuvWQWVxzdePwFfJ7i8vJxzuwCxlMiX/NFYdw=;
        b=EdgBLFRGOi7EFTqXqJF29RaV92f7wTlwbGHhg/jq/N8krp1S++MQXEkatstZVLkZaj
         fCIUkciAdPIU2LW+n4sqgyAwe/8pTKI9saSInraU0i8cLYQAc5db6w8gaYpK/nFu0ngT
         xV+G+Q1V8V/R6KGWEeoi5/DMOcT3NGECo7LVipo8n3X1nIl4jVGpIBOxO5zS9ZDOhzzY
         Av4r7ZCSk2VcPh+fz3/9V/edjrn0SL+9L2x3JAzKO2d1MsmxSG5Hv8BCVz7bd0/hyE09
         5LyDGVZWSkhcyYIzNpkpogq/g/dgM6QStXJkgRjse7JEW3f1MfI0Y1FS+lo1G5oizIhN
         glfQ==
X-Gm-Message-State: AA+aEWbv41h4ujfn/COKP8+YLu2bT1lCJxoQdbiv4O1Bv950uZRdVNgR
        9UDG+9YgnSaCkZsBblwergq0+dnyVcGRdM3Bp7/nomcr3SM/XnzmhQDd2M6fgQOHFDhYp92hXH9
        IfEsksidxdukFBpNKKn+9S0YeWWL49Pvd0jfvQmvvvHDgUDblNWVou2UGAeAJ1PA=
X-Google-Smtp-Source: 
 AFSGD/X8BVJnhX1lmNQUMYxsC8T6iRThe+dyG4HM7k4Gm0hr29XwfgSGiuo7p0CjyK+PdS6bCDI9i1OtyH2JBA==
X-Received: by 2002:a0c:9ad6:: with SMTP id k22mr16628232qvf.7.1544659845334;
 Wed, 12 Dec 2018 16:10:45 -0800 (PST)
Date: Wed, 12 Dec 2018 16:10:26 -0800
Message-Id: <20181213001026.178037-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.20.0.rc2.403.gdbc3b29805-goog
Subject: [PATCH] kvm: vmx: Pass through IA32_TSC_AUX for read iff guest has
 RDTSCP
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Abhiroop Dabral <adabral@paloaltonetworks.com>,
        Jim Mattson <jmattson@google.com>,
        Marc Orr <marcorr@google.com>, Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

If the guest supports RDTSCP, it already has read access to the
hardware IA32_TSC_AUX MSR via RDTSCP, so we can allow it read-access
via RDMSR as well. If the guest doesn't support RDTSCP, then we
should not allow it read access to the hardware IA32_TSC_AUX MSR.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/kvm/vmx.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index c379d0bfdcba9..69deab6f37953 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -6522,6 +6522,10 @@ static void vmx_compute_secondary_exec_control(struct vcpu_vmx *vmx)
 
 	if (vmx_rdtscp_supported()) {
 		bool rdtscp_enabled = guest_cpuid_has(vcpu, X86_FEATURE_RDTSCP);
+
+		vmx_set_intercept_for_msr(vmx->vmcs01.msr_bitmap, MSR_TSC_AUX,
+					  MSR_TYPE_R, !rdtscp_enabled);
+
 		if (!rdtscp_enabled)
 			exec_control &= ~SECONDARY_EXEC_RDTSCP;
 
